pkg/runtime/runtime.go

```diff
@@ -36,48 +40,63 @@ type Config struct {
 	CORS []string
 }
 
-func (r *Runtime) startSyncer(ctx context.Context, syncr sync.ISync) error {
-	if err := r.updateState(ctx, syncr); err != nil {
-		return err
+func (r *Runtime) Start() error {
+	if r.Service == nil {
+		return errors.New("no Service set")
+	}
+	if len(r.SyncImpl) == 0 {
+		return errors.New("no SyncImplementation set")
 	}
+	if r.Evaluator == nil {
+		return errors.New("no Evaluator set")
+	}
+
+	ctx, cancel := signal.NotifyContext(context.Background(), os.Interrupt, syscall.SIGTERM)
+	defer cancel()
 
-	go syncr.Notify(ctx, r.syncNotifier)
-
-	for {
-		select {
-		case <-ctx.Done():
-			return nil
-		case w := <-r.syncNotifier:
-			switch w.GetEvent().EventType {
-			case sync.DefaultEventTypeCreate:
-				r.Logger.Debug("New configuration created")
-				if err := r.updateState(ctx, syncr); err != nil {
-					r.Logger.Error(err.Error())
-				}
-			case sync.DefaultEventTypeModify:
-				r.Logger.Debug("Configuration modified")
-				if err := r.updateState(ctx, syncr); err != nil {
-					r.Logger.Error(err.Error())
-				}
-			case sync.DefaultEventTypeDelete:
-				r.Logger.Debug("Configuration deleted")
-			case sync.DefaultEventTypeReady:
-				r.Logger.Debug("Notifier ready")
+	g, gCtx := errgroup.WithContext(ctx)
+	dataSync := make(chan sync.DataSync)
+
+	// Initialize DataSync channel watcher
+	g.Go(func() error {
+		for {
+			select {
+			case data := <-dataSync:
+				r.updateWithNotify(data)
+			case <-gCtx.Done():
+				return nil
 			}
 		}
+	})
+
+	// Start sync providers
+	for _, s := range r.SyncImpl {
+		p := s
+		g.Go(func() error {
+			return p.Sync(gCtx, dataSync)
+		})
 	}
-}
 
-func (r *Runtime) updateState(ctx context.Context, syncr sync.ISync) error {
-	msg, err := syncr.Fetch(ctx)
-	if err != nil {
-		return fmt.Errorf("fetch: %w", err)
+	g.Go(func() error {
+		return r.Service.Serve(gCtx, r.Evaluator)
+	})
+
+	<-gCtx.Done()
+	if err := g.Wait(); err != nil {
+		return err
 	}
+	return nil
+}
+
+// updateWithNotify helps to update state and notify listeners
+func (r *Runtime) updateWithNotify(data sync.DataSync) {
 	r.mu.Lock()
 	defer r.mu.Unlock()
-	notifications, err := r.Evaluator.SetState(syncr.Source(), msg)
+
+	notifications, err := r.Evaluator.SetState(data.Source, data.FlagData)
 	if err != nil {
-		return fmt.Errorf("set state: %w", err)
+		r.Logger.Error(err.Error())
```

Comment:
it is perhaps useful to log including the `fmt.Errorf("set state: %w", err)` that was there before for additional context

Reply:
This error log should also contain a stack trace, so its no longer necessary, seeing as the log can more easily be traced back to this line

Reply:
If no hard opinion here, I will keep this as it is :)

Reply:
If there is now a stack trace then happy for it to remain as is :)